### ISSUE TITLE
To-be process

### ISSUE THREAD

**Issue by askbulle** [equinor/MAD-VSM-WEB#192]
**User story**
As a user I want to map how a perfect process would look like so that we can communicate around the targets

**Describe the solution you'd like**
- A special case to the version history in future
- Based on current version

**Dependency**
#154 

----
### NOTE: 
- ~~#159 needs to be done in the API before we can start on this task.~~ Labels are now supported in the api.
---- 
# Development breakdown
![image](https://user-images.githubusercontent.com/3164065/135993154-2adcd3c7-f450-4bc9-8020-33799bda2c03.png)

This new feature depends on #159 

**Background**
A `to-be` process is an alternative of the same `original` process, that is mapped how someone imagines a _**perfect**_ ✨ process to be.
The user should be able to quickly change between these two alternatives and compare them together.
The `to-be` process is technically and practically a new process with its own Problems, Questions, Ideas and Risks. Though it shares the same name, labels and access-roles.

### Requirements
- [x] User can flip a toggle to switch between the `original` and the `to-be` process.
- [x] Name is shared
- [x] Access is the same as the original process
- [x] ~Labels are shared between the `original` and the `to-be` processes.~ Depends on #159
- [x] To-be processes should not be shown on the front-page. _(We need some metadata to distinguish them from regular ones...)_
- [x] Individual version history.
- [x] PQIRs are not shared between `original` and `to-be`.
- [x] Disable toggle if `to-be` is not created & you don't have access.
- [x] When duplicating a process, it should duplicate the `current`-process and not create a `To-be`-process.
- [ ] When updating a `To-be` process, it should be reflected in the front-page `last-modified`.

**Use-case**
The API handles all requests related to labels, process-name, and access to the `to-be` process it as if it was the `original` process.

Ex 1.
Client ask for labels for a `to-be` process.
The API answers with the labels for the `original`.

Ex 2
Client adds a label to a `to-be` process.
The API handles it as if it was the `original` process.

----
TODO:
- [ ] Reset Api-endpoint
- [x] Change to-be process toggle
- [x] Update the delete text and dialog design
- [x] Reset button
- [x] Add reset button dialog

**Comment by Magos** [comment]
@askbulle @knudsvik Jeg har diskutert denne og #154 litt med @pagodo, og vi lurer på om dette behovet kan dekkes av #117 - altså at brukerne kan lage en ny versjon av VSMen som er "den vi vil jobbe mot" og vedlike holde disse to ved siden av hverandre uten å involvere versjonshistorikk.
[Metoden for versjonshistorikk vi ser på akkurat nå](https://docs.microsoft.com/en-us/sql/relational-databases/tables/temporal-tables?view=sql-server-ver15) har lett for å assosiere redigering med klokketiden når redigering ble utført, men det å lage framtidige endringer som ikke er gyldige ennå blir mer komplisert.

**Comment by knudsvik** [comment]
Det er en tanke ja, spesielt forståelig for back-end. Kanskje det kunne vært en løsning i front-end å trykke på "Make a to-be process" som i back-end lager en duplikat, mens i front-end vises som en fremtidig versjon?

**Comment by SjoenH** [comment]
Metadata on the original process. Link to the `ToBe`/ideal process
The `ToBe`-process is a new process.

**Comment by SjoenH** [comment]
👋 @knudsvik Thanks for the talk.

We are ready to start on this whenever, add it to our Dev-board (Flyt - issue tracker) to confirm. 👼

**Comment by knudsvik** [comment]
Agreed solution:
- Toggle switch somewhere in the canvas to shift from "current" to "to-be" process (default current)
- First time toggling creates a new blank process and moves to the to-be
- To-be processes should be hidden from front page
- To-be processes should be linked to the "current process" in back-end (new reference column, similar to the dulicateOf)
- To-be processes have their own version history etc just like a regular process

**Comment by SjoenH** [comment]
@Magos mentioned a concern about #159
I guess that both the original process and the to-be process would use the same labels? (Since they really are just two alternatives of the same process.)

**Comment by knudsvik** [comment]
The biggest use-case for labels is to search for them in the front page. The to-be processes are not going to be shown in the front page. But I guess we could just say that the current and the to-be process share labels. Adding to one will add to the other and vice versa

**Comment by SjoenH** [comment]
@Linedueo could you update figma with a simple toggle?

**Comment by Linedueo** [comment]
@SjoenH yup will do

**Comment by Linedueo** [comment]
<img width="1692" alt="Skjermbilde 2021-10-05 kl  09 46 19" src="https://user-images.githubusercontent.com/74361011/135981680-642b2008-76c7-4c95-a66f-c6788eacbaf8.png">
@SjoenH

**Comment by Linedueo** [comment]
https://www.figma.com/file/IkHwmIQrsT0iR34f5R5UnZ/VSM?node-id=2827%3A105422

**Comment by SjoenH** [comment]
@Magos let's have a chat about this.

**Comment by SjoenH** [comment]
@equinor/flyt-product-owners When duplicating a process, does that duplicate include the to-be process?

**Comment by SjoenH** [comment]
@equinor/flyt-product-owners  Should PQIRs be shared between the `To-be` and original process, or should they be separate?

**Comment by knudsvik** [comment]
> @equinor/flyt-product-owners When duplicating a process, does that duplicate include the to-be process?

Nope, not at the first release at least. I've added it to #144 

> @equinor/flyt-product-owners Should PQIRs be shared between the `To-be` and original process, or should they be separate?

No, the `To-be` process should be blank without any cards (except the starting SIPOC cards) when created.

**Comment by Magos** [comment]
Quick check on API design here: Does the front-end need more details about the current/to-be versions than this?
![FgPvuSUeBJ](https://user-images.githubusercontent.com/1004386/137279018-40c0677d-755b-489f-8385-2af9e9e6c63b.png)

This is before adding the mirroring of name, labels and accesses from current to to-be versions. Any comments @SjoenH @martabenthem ?

**Comment by SjoenH** [comment]
A bit unsure about the name `currentProcessId`, but i don't have a better one. So let's go with this.

**Comment by SjoenH** [comment]
So, thinking this through in the front-end.
If no `toBeProcessId`, I can create a new project with `currentProcessId` in the JSON request. The `toBeProcessId` in the `current`-process will then be set by the API.

```json5
{
currentProcessId: theOriginalProjectId,
objects:[...]
}
```
For the toggle-state-logic...
I can indicate on the `toBe`-process that we are currently showing that one by checking that the `currentProcessId` field is not empty.

If none of them are set, we are showing the `current` process: and would need to create the `toBe` process when toggling it.
If only the `toBeProcessId` is set, then we are also on the `current`-process, but can just navigate to that one when toggling.

**Comment by Magos** [comment]
@ntorsen The new version with FromSqlRaw and parameters (should deal with injection attacks) is deployed to Dev & Test now.

**Comment by SjoenH** [comment]
Still need to deploy front-end to test

**Comment by SjoenH** [comment]
To-do:
Back-end:
- [ ] Guard post endpoint against creating empty projects
Front-end:
- [ ] Hide toggle if `to-be` is not created.

**Comment by Magos** [comment]
Empty to-be processes should not be creatable by users who can't edit the `current` version in the current Dev/Test environments.

**Comment by SjoenH** [comment]
@ntorsen @Linedueo ready for testing 💪 
https://web-vsm-test.radix.equinor.com

**Comment by ntorsen** [comment]
@SjoenH I get an error when clicking on the 'to-be process'
![image](https://user-images.githubusercontent.com/89014017/139803014-e39c14af-88b1-48d0-b206-4845f017b208.png)

**Comment by SjoenH** [comment]
@ntorsen This was caused when clicking the toggle before @Magos did his fix.
Please test with a new process.

**Comment by ntorsen** [comment]
To-be process has been tested in Test env, and it looks good 👍 

@Magos when I deleted a process I get an error code 400. So it doesn't look like it is deleted correctly. 
![image](https://user-images.githubusercontent.com/89014017/139816651-c0a3bb8f-d2aa-4870-b589-1a1dd01439c3.png)

When searching for the process after it has been 'deleted', I get this
![image](https://user-images.githubusercontent.com/89014017/139817078-8b810838-6735-4e41-b05e-c574238a17c9.png)

**Comment by Magos** [comment]
I'm now putting out [Release 188](https://dev.azure.com/statoil-mad/VSM/_apps/hub/ms.vss-releaseManagement-web.cd-release-progress?_a=release-pipeline-progress&releaseId=188), with automatic deletion of to-be processes if the current process is deleted. This should prevent that error.
Note that this is not symmetrical - deleting a to-be process should not delete its current version automatically.

**Comment by SjoenH** [comment]
@Magos 
The GUI doesn't differentiate between renaming or managing access on a `To-be` process or a `current` process.
So, wouldn't it make sense that deleting works the same as well? 🤔

**Comment by SjoenH** [comment]
Considering there's always just one process from the user-perspective. We are abstracting away the underlying two processes...
I don't see why we should make it possible to delete a To-be process, and leave the current.
The UI doesn't differentiate today, so we would need to make that clear somehow if we want this feature. 

I'm open to hear your points @Magos. 
I'm leaning towards that the API deletes both process variations if one of them are deleted.

**Comment by Magos** [comment]
I was a little unsure myself, so I ended up leaving it at the asymmetric version, but symmetrically deleting both if the user tries to delete either makes sense too. I'm deploying [that version](https://dev.azure.com/statoil-mad/VSM/_releaseProgress?_a=release-pipeline-progress&releaseId=189) out now.

**Comment by SjoenH** [comment]
Super. Thank you

**Comment by SjoenH** [comment]
@ntorsen I believe this is ready for testing again.

**Comment by SjoenH** [comment]
Need to look over the duplicate feature... 🙃 

To-do Front-end
- [ ] When duplicating a process, it should duplicate the `current`-process and not create a `To-be`-process.

**Comment by SjoenH** [comment]
@ntorsen ready for testing.
I fixed the point above.

**Comment by ntorsen** [comment]
Tried duplicating while on current-process and on to-be process, both gives a duplicate of the current. 
Also I've done an overall testing of the other points mentioned at the top- everything works good 😎

**Comment by SjoenH** [comment]
### Ready for testing in **QA** @knudsvik & @askbulle.

@Linedueo I've not heard anything from you about this task, but I have [tagged you earlier](https://github.com/equinor/MAD-VSM-WEB/issues/192#issuecomment-957125059), so you should have been notified as agreed upon... So I'm assuming everything is OK? 😅

I'm moving front-end into QA. @Magos I believe the API is already in QA? If not, please release it when you have time. 😄

**Comment by knudsvik** [comment]
When I click the To-Be button (from process 129), a new Untitled process is created, not the To-Be process, so something is off...

**Comment by Magos** [comment]
[Deployet To-be-process backend til QA nå](https://dev.azure.com/statoil-mad/VSM/_releaseProgress?_a=release-environment-logs&releaseId=189&environmentId=755), og min egentest ser bra ut på web-vsm-qa.radix.equinor.com.

**Comment by SjoenH** [comment]
@knudsvik Sorry about that. Would you try again now?

**Comment by knudsvik** [comment]
## Results from initial QA:
- The to-be switch seems a bit counter-intuitive when it's a on-off switch. Instead, would it make more sense to have a switch between two options (`Current` and `To-be`)?
- Deleting a process when in To-be mode should only delete the To-be process, or we would risk a lot of angry mappers. How can we fix that problem? Detailing the text in the popup message better as well? or maybe the popup message should have two buttons, one for deleting Current and To-be and another for only deleting the To-be?
-

**Comment by SjoenH** [comment]
@knudsvik Thanks for testing.
@Linedueo we would need some input on best-practice behaviour here.

**Comment by Linedueo** [comment]
@knudsvik The to be switch could be solved by using a toggle button instead of a switch. It would look like this:
<img width="613" alt="Skjermbilde 2021-11-04 kl  09 50 46" src="https://user-images.githubusercontent.com/74361011/140284725-b5f9eeb4-c3b4-462e-8b5b-df2be25c0016.png">
<img width="1094" alt="Skjermbilde 2021-11-04 kl  09 51 43" src="https://user-images.githubusercontent.com/74361011/140284865-e4f0d53a-183a-48c1-be90-55c44a5fcbac.png">

**Comment by knudsvik** [comment]
> @knudsvik The to be switch could be solved by using a toggle button instead of a switch. 

Looked way better :) Yes please

**Comment by askbulle** [comment]
@Linedueo I love the toggle button

**Comment by askbulle** [comment]
@SjoenH Re: Why someone would like to delete only To-Be

This is mostly speculation. Our main problem is that too many processes are unmapped all together. Flyt makes it extremely easy to start mapping - just open the tool and new process is saved even if just looking/browsing what the tool looks like. Same for To-be.  

This is good for mapping but causes bad digital hygiene wrt metrics. One very interesting metric to have for follow up of active process improvement would be to visualize # prosesses with To-be process. Having hordes of "someone clicked on toggle button"-processes makes it harder. 

For actual front runner users I'd say that it is early to say why, but my hunch is that we will have users who wish to delete. And is it not odd to not be able to? Not being able to delete sounds disturbing in general. (That would be like a family checking into a hotel and it being impossible for single family members to leave if not entire family dies...)

**Comment by knudsvik** [comment]
A user may want to delete the to-be process and start from scratch. But the most important thing in my mind is to secure that a user do not delete both current and to-be by a mistake thinking that he/she will only delete the to-be...

**Comment by SjoenH** [comment]
We could change the button to say `Delete To-be process` when displaying that one.
But we would lose that data+version-history when they delete it. 
If that's not a concern, we may do this change quite quickly.

**Comment by SjoenH** [comment]
@askbulle  (+ @knudsvik)
We are able to differentiate a `To-be`-process from a `current`-process. But it has to be accounted for with regard to metrics.


As it is now: (as mentioned in https://github.com/equinor/MAD-VSM-WEB/issues/192#issuecomment-943126907)
- If it is a `To-be` process, it should include a number in the field `currentProcessId`.
- If not, it's a `current`-process. (The `currentProcessId`-field is empty if it is a `current`-process)

| Process-type | `currentProcessId`-field |
|--------------|---------------------|
| Current | `null` |
| To-be | number linking back to the `current`-process|

**Comment by SjoenH** [comment]
@Linedueo do you have any feedback on this?

**Comment by Linedueo** [comment]
@SjoenH @knudsvik would it be possible to delete "to be" process only when in the "to be" process. And when in current you delete the entire process

**Comment by knudsvik** [comment]
@Linedueo That's what @SjoenH proposed as well I think? I'm somehow ambivalent to it. I trust @Linedueo to come up with the best suggestion.

Something needs to be done with the dialogue window anyhow to explain better what is about to happen (and add info that it is not possible to recover it once it's deleted).

**Comment by knudsvik** [comment]
@SjoenH Can you add last updated/modified update when To-be process has changes as a requirement? Ref #258

**Comment by SjoenH** [comment]
Sure, 
Added another requirement:
"When updating a To-be process, it should be reflected in the front-page last-modified." #282

**Comment by Linedueo** [comment]
<img width="1905" alt="Skjermbilde 2021-11-10 kl  13 50 43" src="https://user-images.githubusercontent.com/74361011/141116108-3ca32ed7-f308-4f35-94a4-71d0fdd228c0.png">
@SjoenH @knudsvik @askbulle I´ve added som screenshots here of how we could solve the "reset to be version" and "delete VSM". Ive added a reset page button in the to be version. A dialog will pop up with an explanatory text (need help with this). And once you choose to delete an entire VSM you will be aware of that it will remove current and to be version.

**Comment by Linedueo** [comment]
link: https://www.figma.com/file/IkHwmIQrsT0iR34f5R5UnZ/VSM?node-id=2994%3A117375

**Comment by askbulle** [comment]
@Linedueo @SjoenH 

Thorjan and I both like the design for both delete and reset. Explanatory texts
Delete:
Are you sure you want to delete this process? By doing so you will delete all versions of Current and To-be processes, neither of which will be recoverable.

Reset:
Are you sure you want to reset the To-Be process? By doing so you will reset to a blank To-Be process and previous version of the To-be process will not be recoverable

**Comment by SjoenH** [comment]
Some thoughts...
P: People don't read text.
S: We need intuitive design... or at least the possibility to undo actions.

I: We could **archive instead of delete** and make it possible to unarchive it.
C: Then it isn't a big deal if someone archives the whole process by accident.

Q: What then if you really want to delete something?
A: Maybe only show the delete-button for archived processes?

Q: What about the front-page search?
A: Archived processes could be returned, but would need to be displayed as archived. Either on the card itself or in another section etc.

**About the reset button**
Q: How often would the reset button be used? Do we want to encourage frequent resetting of the `to-be`-view?

Q: Why only for the `to-be` process?

**Comment by Linedueo** [comment]
I did add a snackbar to the design so that you can undo actions
<img width="794" alt="Skjermbilde 2021-11-11 kl  11 44 52" src="https://user-images.githubusercontent.com/74361011/141284809-63974165-8c95-49a0-9c62-0c12e5ecc501.png">

**Comment by SjoenH** [comment]
> I did add a snackbar to the design so that you can undo actions
> 
> <img width="794" alt="Skjermbilde 2021-11-11 kl  11 44 52" src="https://user-images.githubusercontent.com/74361011/141284809-63974165-8c95-49a0-9c62-0c12e5ecc501.png">
> 
> 

That design would need a lot of thought put into the implementation. So I would suggest we drop it at this point. We are already showing a prompt before they are doing the change. 

We may consider adding it after versioning is available.

**Comment by knudsvik** [comment]
> That design would need a lot of thought put into the implementation. So I would suggest we drop it at this point. We are already showing a prompt before they are doing the change.
> 
> We may consider adding it after versioning is available.

Agreed, but love the idea though! Made it into a feature #295

**Comment by Linedueo** [comment]
👍🏻

**Comment by SjoenH** [comment]
@Magos Let's have a quick talk about this one.

**Comment by SjoenH** [comment]
**Regarding resetting a to-be process.**
👋 @knudsvik - After talk with @Magos we are leaning towards keeping the PQIRs, only unlinking them before resetting the process.
I believe it to be safer with regard to versioning in the future.

Also, seems like a reset-endpoint in the API might be better than doing it in the Front-end-client. The reason is that this is business logic concerning the database implementation details.

We could also consider making a create-endpoint. But not needed atm.
---
@Magos 
The structure for creating a new project today looks like this:
```json5
{
    objects: [
      {
        parent: 0,
        fkObjectType: vsmObjectTypes.process,
        childObjects: [
          { fkObjectType: vsmObjectTypes.supplier },
          { fkObjectType: vsmObjectTypes.input },
          { fkObjectType: vsmObjectTypes.mainActivity },
          { fkObjectType: vsmObjectTypes.output },
          { fkObjectType: vsmObjectTypes.customer },
        ],
      },
    ],
  }

```
And my enum look like this
```ts
export enum vsmObjectTypes {
  process = 1,
  supplier,
  input,
  mainActivity,
  subActivity,
  text,
  waiting,
  output,
  customer,
  choice,
}

```

**Comment by Linedueo** [comment]
<img width="600" alt="Skjermbilde 2021-11-16 kl  20 44 11" src="https://user-images.githubusercontent.com/74361011/142054732-c7a7876c-0105-4910-8807-1bff8c10e5fe.png">


@SjoenH Let´s treat it as a regular contained button in disabled state

**Comment by SjoenH** [comment]
Thanks. Looks much better

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by SjoenH** [comment]
Front-end done. Waiting on API endpoint. @Magos

**Comment by Magos** [comment]
Reset endpoint is now building to Dev environment. Example URL:
`PATCH /api/v1/Project/776/Reset`
Currently, it's available for roles {Owner, Admin}.

**Comment by SjoenH** [comment]
@Magos I get 404 in dev-env. 
For example: `PATCH | https://api.statoil.com/app/vsm/dev/api/v1.0/project/777/reset`

**Comment by SjoenH** [comment]
Works good now @Magos 💪 
Would you release to Test-env?

**Comment by SjoenH** [comment]
@ntorsen Ready for testing 💪

**Comment by ntorsen** [comment]
### **Results from testing:**
User can flip a toggle to switch between the original and the to-be process. ✔️
Name is shared ✔️
Access is the same as the original process ✔️
To-be processes should not be shown on the front-page. (We need some metadata to distinguish them from regular ones...) ✔️
Individual version history. ✔️
PQIRs are not shared between original and to-be. ✔️
Disable toggle if to-be is not created & you don't have access. ✔️
When duplicating a process, it should duplicate the current-process and not create a To-be-process. ✔️
**When updating a To-be process, it should be reflected in the front-page last-modified.** ⛔ 

It is not reflected in the front-page last modified when I update a To-be process. 
Also, if I create a new process and make a to-be process this does not show in "See my processes". Newly created processes land on that page, but if you do something on the To-be they will disappear from this page. 
![image](https://user-images.githubusercontent.com/89014017/143014288-f5e673ee-4496-4643-9dea-44672e824673.png)

No problem in setting a process as Favorite after making a To-be.

**Comment by Magos** [comment]
>When updating a To-be process, it should be reflected in the front-page last-modified. ⛔

On this: The approach I'm planning for #154 means we will soon abandon the ChangeLog-based way this is updated now. As such, I'm inclined not to change this now, but instead when we release versioning.

**Comment by SjoenH** [comment]
Moving this to in-progress since we need to fix
- [ ] #307

**Comment by SjoenH** [comment]
Let's get this into QA. @Magos 💪  [Ref](https://github.com/equinor/MAD-VSM-WEB/issues/307#issuecomment-976653205)

**Comment by Magos** [comment]
Building to QA environment now.

**Comment by SjoenH** [comment]
@askbulle @knudsvik @Linedueo @ntorsen Ready for testing in QA.
https://web-vsm-qa.radix.equinor.com/

Let us know of any concerns.

**Comment by ntorsen** [comment]
No concerns here. Working good in QA 👍

**Comment by SjoenH** [comment]
Super. Waiting on PO approval.

**Comment by knudsvik** [comment]
Looks good, you may release 💪

**Comment by SjoenH** [comment]
This is now released together with #254 to Prod :tada:
